core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -584,16 +584,21 @@ public void runInContext() {
 
     @Override
     public void closed(final Status status) {
+      // For cancellations, promptly inform any users of the context that their work should be
+      // aborted. Otherwise, we can wait until pending work is done.
+      if (!status.isOk()) {
+        context.cancel(status.getCause());
```

Comment:
Since the context listeners may be using directExecutor, I'd like to avoid calling them from the network thread. Can you schedule this on `executor`? (not callExecutor)

@zhangkun83, since we've talked so much about which executor to use when, how does this sound to you? Using `callExecutor` is known-problematic, and not using any executor makes me a bit worried.